### PR TITLE
New Resource: `azurerm_spring_cloud_connection` and `azurerm_app_service_connection`

### PR DESCRIPTION
**Pull request by jiaweitao001**
* Add a new resource called `service connector`.  It helps connect Azure compute service to other backing services. 

* Official documentation: https://docs.microsoft.com/en-us/azure/service-connector/

### CODE REVIEW

File: internal/services/serviceconnector/sdk/2022-05-01/links/client.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package links
```
Comment (tombuildsstuff) [inline internal/services/serviceconnector/sdk/2022-05-01/links/client.go:1]: @jiaweitao001 where's this SDK come from? is there an upstream PR to add support for it?

Comment (jiaweitao001) [reply internal/services/serviceconnector/sdk/2022-05-01/links/client.go:1]: Hi @tombuildsstuff, this SDK was generated by pandora sdk tool.  There's no upstream PR.

**Comment by jiaweitao001** [comment]
Hi @tombuildsstuff , the upstream supporting PR has been added. This one is ready for review. Thanks!

**Review by katbyte — CHANGES_REQUESTED** [review]
@jiaweitao001 - can we update this to use the now public hashicorp/go-azure-sdk?

File: website/docs/r/service_connector_app_service.html.markdown [hunk 2]
```diff
@@ -0,0 +1,133 @@
+---
+subcategory: "App Service (Web Apps)"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_app_service_connection"
+description: |-
+  Manages a service connector for app service.
```
Comment (katbyte) [inline website/docs/r/service_connector_app_service.html.markdown:6]: this doesn't match the PR title?

File: internal/services/serviceconnector/sdk/2022-05-01/links/client.go (hunk 1 above)
Comment (katbyte) [inline internal/services/serviceconnector/sdk/2022-05-01/links/client.go:1]: Where is this sdk comming from? we should only be using the go-azure-sdk one?

File: internal/services/privatedns/parse/a_record.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package parse
```
Comment (katbyte) [inline internal/services/privatedns/parse/a_record.go:1]: why are we modifying the private dns service?

**Review by katbyte — CHANGES_REQUESTED** [review]
comments left inline

File: website/docs/r/service_connector_app_service.html.markdown (hunk 2 above)
Comment (jiaweitao001) [reply website/docs/r/service_connector_app_service.html.markdown:6]: Actually this PR contains 2 new resources. PR title modified.

File: internal/services/serviceconnector/sdk/2022-05-01/links/client.go (hunk 1 above)
Comment (jiaweitao001) [reply internal/services/serviceconnector/sdk/2022-05-01/links/client.go]: This is the old sdk, removed.

File: internal/services/privatedns/parse/a_record.go (hunk 3 above)
Comment (jiaweitao001) [reply internal/services/privatedns/parse/a_record.go]: Should not be included here. Removed.

File: internal/services/serviceconnector/service_connector_app_service_resource.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+		"vnet_solution": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				string(servicelinker.VNetSolutionTypeServiceEndpoint),
+				string(servicelinker.VNetSolutionTypePrivateLink),
+			}, false),
+		},
+
+		"auth_info": authInfoSchema(),
```
Comment (katbyte) [inline internal/services/serviceconnector/service_connector_app_service_resource.go:80]: info is redundant 
```suggestion
		"authentication": authInfoSchema(),
```

**Comment by jiaweitao001** [comment]
Per discussion with the service team, this feature is never released. Will withdraw the PR.

File: website/docs/r/service_connector_app_service.html.markdown [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the service connection. Changing this forces a new resource to be created.
+
+* `app_service_id` - (Required) The ID of the data source web app. Changing this forces a new resource to be created.
+
+* `target_resource_id` - (Required) The ID of the target resource. Changing this forces a new resource to be created. Possible values are `Postgres`, `PostgresFlexible`, `Mysql`, `Sql`, `Redis`, `RedisEnterprise`, `CosmosCassandra`, `CosmosGremlin`, `CosmosMongo`, `CosmosSql`, `CosmosTable`, `StorageBlob`, `StorageQueue`, `StorageFile`, `StorageTable`, `AppConfig`, `EventHub`, `ServiceBus`, `SignalR`, `WebPubSub`, `ConfluentKafka`.
+
+* `auth_info` - (Required) The authentication info. An `auth_info` block as defined below.
```
Comment (katbyte) [inline website/docs/r/service_connector_app_service.html.markdown:91]: this wasn't renamed to authentication?

File: website/docs/r/service_connector_spring_cloud.html.markdown [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the service connection. Changing this forces a new resource to be created.
+
+* `spring_cloud_id` - (Required) The ID of the data source spring cloud. Changing this forces a new resource to be created.
+
+* `target_resource_id` - (Required) The ID of the target resource. Changing this forces a new resource to be created. Possible values are `Postgres`, `PostgresFlexible`, `Mysql`, `Sql`, `Redis`, `RedisEnterprise`, `CosmosCassandra`, `CosmosGremlin`, `CosmosMongo`, `CosmosSql`, `CosmosTable`, `StorageBlob`, `StorageQueue`, `StorageFile`, `StorageTable`, `AppConfig`, `EventHub`, `ServiceBus`, `SignalR`, `WebPubSub`, `ConfluentKafka`.
+
+* `auth_info` - (Required) The authentication info. An `auth_info` block as defined below.
```
Comment (katbyte) [inline website/docs/r/service_connector_spring_cloud.html.markdown:96]: same here

**Review by katbyte — APPROVED** [review]
two comments to address but otherwise LGTM 🌳

File: website/docs/r/service_connector_app_service.html.markdown (hunk 5 above)
Comment (jiaweitao001) [reply website/docs/r/service_connector_app_service.html.markdown]: Renamed. Thanks.

File: website/docs/r/service_connector_spring_cloud.html.markdown (hunk 6 above)
Comment (jiaweitao001) [reply website/docs/r/service_connector_spring_cloud.html.markdown]: Renamed.

**Comment by magodo** [comment]
@jiaweitao001 The tests on TC are failing:

```
Test Failed
------- Stdout: -------
=== RUN   TestAccServiceConnectorAppServiceCosmosdb_basic
=== PAUSE TestAccServiceConnectorAppServiceCosmosdb_basic
=== CONT  TestAccServiceConnectorAppServiceCosmosdb_basic
=== CONT  TestAccServiceConnectorAppServiceCosmosdb_basic
testcase.go:110: Step 1/2 error: After applying this test step, the plan was not empty.
stdout:
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
~ update in-place
Terraform will perform the following actions:
# azurerm_app_service_connection.test will be updated in-place
~ resource "azurerm_app_service_connection" "test" {
- client_type        = "none" -> null
id                 = "/subscriptions/*******/resourceGroups/acctestRG-220826014416084428/providers/Microsoft.Web/sites/linuxwebapplvrfw/providers/Microsoft.ServiceLinker/linkers/acctestserviceconnector220826014416084428"
name               = "acctestserviceconnector220826014416084428"
# (2 unchanged attributes hidden)
# (1 unchanged block hidden)
}
Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccServiceConnectorAppServiceCosmosdb_basic (802.46s)
```

(same for the others)

**Comment by jiaweitao001** [comment]
Hi @magodo , failing tests are now fixed. Thanks.

```
--- PASS: TestAccServiceConnectorAppServiceCosmosdb_basic (889.80s)
--- PASS: TestAccServiceConnectorAppService_complete (945.70s)
--- PASS: TestAccServiceConnectorAppServiceCosmosdb_update (1150.62s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/serviceconnector      1164.093s


--- PASS: TestAccServiceConnectorSpringCloudCosmosdb_basic (1110.27s)
--- PASS: TestAccServiceConnectorSpringCloud_complete (1152.31s)
--- PASS: TestAccServiceConnectorSpringCloudCosmosdb_update (1549.22s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/serviceconnector      1565.813s
```

**Comment by magodo** [comment]
@jiaweitao001 The TC is still failing for the app service one:

```
Test Failed
------- Stdout: -------
=== RUN   TestAccServiceConnectorAppServiceCosmosdb_basic
=== PAUSE TestAccServiceConnectorAppServiceCosmosdb_basic
=== CONT  TestAccServiceConnectorAppServiceCosmosdb_basic
=== CONT  TestAccServiceConnectorAppServiceCosmosdb_basic
testcase.go:110: Step 1/2 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
stdout
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
~ update in-place
Terraform will perform the following actions:
# azurerm_linux_web_app.test will be updated in-place
~ resource "azurerm_linux_web_app" "test" {
~ app_settings                      = {
- "AZURE_COSMOS_LISTCONNECTIONSTRINGURL" = "https://management.azure.com/subscriptions/*******/resourceGroups/acctestRG-220826073341929034/providers/Microsoft.DocumentDB/databaseAccounts/acctestacche1ln/listConnectionStrings?api-version=2021-04-15" -> null
- "AZURE_COSMOS_RESOURCEENDPOINT"        = "[https://acctestacche1ln.documents.azure.com:443/](https://acctestacche1ln.documents.azure.com/)" -> null
- "AZURE_COSMOS_SCOPE"                   = "https://management.azure.com/.default" -> null
}
id                                = "/subscriptions/*******/resourceGroups/acctestRG-220826073341929034/providers/Microsoft.Web/sites/linuxwebapphe1ln"
name                              = "linuxwebapphe1ln"
tags                              = {}
# (17 unchanged attributes hidden)
- identity {
- identity_ids = [] -> null
- principal_id = "a0d1cfe3-2808-4d6f-8ab3-fd4e7966d370" -> null
- tenant_id    = "*******" -> null
- type         = "SystemAssigned" -> null
}
# (2 unchanged blocks hidden)
}
Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccServiceConnectorAppServiceCosmosdb_basic (875.59s)
```

**Comment by jiaweitao001** [comment]
Hi @magodo , the acc tests are now fixed in pipeline.

**Comment by magodo** [comment]
image

**Pull request merged**